### ISSUE TITLE
do not show [Process exited] in finished terminals

### ISSUE THREAD

**Issue by Hubro** [neovim/neovim#14986]
The `[Process exited]` prompt is extremely handy when running non-interactive commands like `make` or `make test`, as it lets you examine the output before the window disappears.

However, for running interactive applications like `lazygit` or `htop`, it is just an additional keystroke you need to press to close the window.

I have this keybinding set up to launch `lazygit` in a new tab:

```vim
" Open lazygit in a new tab
if has("nvim")
    nnoremap <Leader>g :tabnew<CR>:terminal lazygit<CR>i
endif
```

When I'm done I press `q`, which exits `lazygit`. But then I have to press another key to close the window. This isn't a huge deal of course, but it does irk me since it's a completely unnecessary keypress. I will never need to examine the terminal output after `lazygit` exits, so it would feel a lot nicer if I could simply press `q` and go right back to the file I was editing.

The same is true when I just want to open a quick terminal to examine something in the command line. When I press <kbd>Ctrl</kbd><kbd>d</kbd> to exit the terminal, it feels unnecessary to have to press another key to close the window.

I would love an option when opening a new terminal to not prompt for a keypress after the command exits.

Maybe:

```vim
:terminal <noprompt> lazygit
```

---

I tried hacking together a workaround using `TermClose`, but it seems that event fires before the prompt is displayed, so I'm not sure if it can work:

```
autocmd TermClose term://*lazygit <CR>
```

![image](https://user-images.githubusercontent.com/597206/124392950-676b0880-dcf8-11eb-812f-32c6b59d1551.png)

**Comment by jose-elias-alvarez** [comment]
I have a similar setup (specifically with Lazygit too), and I think if you modify your autocommand, it'll work the way you want:

```vim
autocmd TermClose term://*lazygit lua vim.api.nvim_input("<CR>")
```

**Comment by justinmk** [comment]
we absolutely should not add an "option" or other special case thing for this. The message really shouldn't be there in the first place, we should remove it and indicate the final status of the job in some other way.

**Comment by oblitum** [comment]
Vim provides ++close for this: `:terminal ++close <command>` (e.g. `:terminal ++close true`). See [`:h term++close`](https://vimhelp.org/terminal.txt.html#term%2B%2Bclose).

**Comment by justinmk** [comment]
With https://github.com/neovim/neovim/pull/15406 this can be achieved by:

    autocmd TermClose * if !v:event.status | bdelete | endif

**Comment by oblitum** [comment]
@justinmk it doesn't. None of this works because the target terminal buffer may not be in focus on `TermClose`. I tried to set `autocmd TermClose * call nvim_input('<CR>')` and it sent an <kbd>enter</kbd> to an unrelated buffer instead of the terminal one. This happened after closing nvim-dap-ui, which closes all its buffers on `.close()`, except for the terminal:

![nvim-dap-ui](https://user-images.githubusercontent.com/1269815/130146808-419ad5ec-f66c-462f-879b-878e149a8209.gif)

**Comment by gpanders** [comment]
> @justinmk it doesn't. None of this works because the target terminal buffer may not be in focus on `TermClose`. I tried to set `autocmd TermClose * call nvim_input('<CR>')` and it sent an enter to an unrelated buffer instead of the terminal one, `autocmd TermClose * quit` closed _just_ the other non-terminal buffer, etc.

Did you try what @justinmk suggested and use `bdelete!`? It's no surprise that `call nvim_input('<CR>')` didn't work, as the autocommand does not automatically focus the terminal buffer before executing the command. `bdelete` should work just fine though, as all buffer commands do execute in the context of the autocommand buffer.

I don't think we should be adding any options to `:terminal`. We should instead just create better defaults. I think a good start would be to automatically close the terminal buffer if the user did not provide a `{cmd}` to `:terminal` **AND** the terminal exits with status code 0 (no errors).

I think this would be pretty easy to implement once #15406 is merged.

**Comment by oblitum** [comment]
@gpanders I did, and as expected, `:bdelete` alone wouldn't help, only `:bdelete!`, but as you can notice, I even tried `:quit` as well. None works.

**Comment by gpanders** [comment]
> @gpanders I did, and as expected, `:bdelete` alone wouldn't help, only `:bdelete!`, but as you can notice, I even tried `:quit` as well. None works.

This is a known issue and will be fixed in https://github.com/neovim/neovim/pull/15402. `:bdelete!` does work, so I'm not sure why you're saying it doesn't.

**Comment by oblitum** [comment]
I'm on 0.5 and it just closes the original buffer and leave the disposed terminal one still open.

**Comment by gpanders** [comment]
This looks like an issue with the DAP plugin. We're getting off-topic of the issue at hand, so please open a new issue here or with nvim-dap.

**Comment by oblitum** [comment]
@gpanders that doesn't make sense, I expect an `autocmd TermClose * bdelete!` to target the terminal buffer that trigger the `TermClose`, regardless from where this terminal buffer originates.

**Comment by oblitum** [comment]
From what I can tell, this this solution, to get something possibly reliable in any case, I'd need to involve a `:bufdo` and apply an `if !v:event.status`, or something in that sense, which is overkill.

**Comment by oblitum** [comment]
I found a way to make this work properly in current state of matters:

```vim
:autocmd TermClose * execute 'bdelete! ' . expand('<abuf>')
```

**Comment by zeertzjq** [comment]
Note: Because of #16257, this solution leaks memory with every terminal closed with `:bdelete` on `TermClose`.

**Comment by zeertzjq** [comment]
> Note: Because of #16257, the `:bdelete` solution leaks memory with every terminal closed with `:bdelete` on `TermClose`.

Fixed in #16264.

**Comment by sisrfeng** [comment]
*Not so good*:
![image](https://user-images.githubusercontent.com/53520949/194736128-ee6b7a30-8d70-43f3-bccb-f5187343d05c.png)



From https://github.com/voldikss/vim-floaterm/issues/315:
> ... to auto delete terminal buffers when they exit. This isn't working great with vim-floaterm because the terminal gets delete but the floating window is left behind empty.

Similar problem with the plugin
```vim
plug 'https://github.com/sisrfeng/sk'
    plug 'https://github.com/sisrfeng/skim' ,
          \ {
           \ 'dir': $LEO_TOOL . '/skim',
           \ 'do': './install',
          \ }
```

![image](https://user-images.githubusercontent.com/53520949/194685247-26c1a05d-54a9-477b-85f9-5e69333aafcc.png)

**Comment by mikesmithgh** [comment]
Here is my workaround to deleting that pesky line 😂

```lua
  vim.api.nvim_create_autocmd('TermClose', {
    group = vim.api.nvim_create_augroup('DeleteLastLineTermClose', { clear = true }),
    pattern = { '*' },
    callback = function(ev)
      local delete_line_timer = vim.fn.timer_start(
        100,
        function(t) ---@diagnostic disable-line: redundant-parameter
          local process_exited_line = vim.fn.search('\\[process exited \\d\\]', 'bn')
          if process_exited_line > 0 then
            vim.api.nvim_set_option_value('modifiable', true, { buf = ev.buf, })
            vim.api.nvim_buf_set_lines(ev.buf, process_exited_line - 1, process_exited_line, true, { '' })
            vim.api.nvim_set_option_value('modifiable', false, { buf = ev.buf, })
            vim.fn.timer_stop(t)
          end
        end,
        { ['repeat'] = -1 } -- repeat indefinitely but will be cancelled after 3 seconds
      )

      -- give at most 3 seconds of an attempt to delete the line
      vim.defer_fn(
        function() vim.fn.timer_stop(delete_line_timer) end,
        3000
      )
    end
  })
```
Hope this helps someone!

**Comment by sisrfeng** [comment]
`tno <M-C-D>   <C-\><C-N>:bd!<cr>`  solves my problem.  (But if your using neovide, `Alt` will not work untill next release.)

**Comment by tmccombs** [comment]
I would like a way to keep the terminal open, but don't show the [Process exited] line. I think that is what [mikesmithgh](https://github.com/mikesmithgh)'s code does, although that is definitely kind of hacky.

**Comment by mikesmithgh** [comment]
> I would like a way to keep the terminal open, but don't show the [Process exited] line. I think that is what [mikesmithgh](https://github.com/mikesmithgh)'s code does, although that is definitely kind of hacky.

Yep that is what mine is doing. 👍

**Comment by justinmk** [comment]
With #15440 , `:terminal` is closed if the command exits without error. [:help default-autocmds](https://neovim.io/doc/user/vim_diff.html#default-autocmds)

To disable this behavior:

    :autocmd! nvim_terminal TermClose

## Status

But we still need to remove the `[Process exited]` text from the buffer contents (show it in 'statusline' instead, like Vim).

## Proposal 1

- The main missing piece is a way to get the result (error code) of the terminal process.
    - Add a `code` field to the result of `nvim_get_chan_info`.
    - ⚠️ Currently, when a process exits, Nvim destroys the job info, so `nvim_get_chan_info()` won't work.
        - To fix that, the _minimum_ interface is to change the internal channels/jobs list to keep ring buffer of the last N channels, instead of destroying the channel immediately.
        - Deprecate `nvim_list_chans`.
        - Deprecate `nvim_get_chan_info`.
        - Introduce `nvim_chan_get(channelId, opts): List<Dict>` 
- The default 'statusline' should show the exit code (e.g. `Exit code: 0`). Users with a custom statusline can get the exit code via `nvim_chan_get()`.

## Proposal 2

Allow job handlers to be defined on existing jobs: https://github.com/neovim/neovim/issues/5023 . Then the exit-code can be captured by a Lua callback which sets the statusline or whatever.

**Comment by GitMurf** [comment]
In case anyone is running into this still today, here is my solution (based off some of the solutions from the past above). If any of the maintainers / experts have any advice for me of how to make this better or if there is another solution, please let me know! Thanks!

You will notice I added a vim.schedule(). For some reason without pushing the bdelete to the next event loop cycle, the terminal buffer was still showing up in `ls!` buffer list. This fixed it.

```lua
vim.api.nvim_create_autocmd('TermClose', {
  pattern = '*',
  callback = function()
    vim.schedule(function()
      if vim.bo.buftype == 'terminal' and vim.v.shell_error == 0 then
        vim.cmd('bdelete! ' .. vim.fn.expand('<abuf>'))
      end
    end)
  end,
})
```

**Comment by justinmk** [comment]
# Resolution

With #15440 , `:terminal` is closed if the command exits without error. [:help default-autocmds](https://neovim.io/doc/user/vim_diff.html#default-autocmds)

To disable this behavior:

    :autocmd! nvim_terminal TermClose

With https://github.com/neovim/neovim/pull/37987 , 

- `[Process exited]` is presented as virtualtext instead of modifying the buffer contents
- also displayed in default 'statusline'

## Proposal 1 [DONE]

- [x] Need a way to get the result (error code) of the terminal process.
    - Add a `exitcode` field to the result of `nvim_get_chan_info`.
- [x] The default 'statusline' should show the exit code (e.g. `Exit code: 0`). Users with a custom statusline can get the exit code via `nvim_get_chan_info()`.

## Proposal 2 [TODO]

- Allow job handlers to be defined on existing jobs: https://github.com/neovim/neovim/issues/5023 . Then the exit-code can be captured by a Lua callback which sets the statusline or whatever.
- [TBD] ⚠️ When a process exits, Nvim destroys the job info, so `nvim_get_chan_info()` won't work.
    - To fix that, the _minimum_ interface is to change the internal channels/jobs list to keep ring buffer of the last N channels, instead of destroying the channel immediately.
    - Deprecate `nvim_list_chans`.
    - Deprecate `nvim_get_chan_info`.
    - Introduce `nvim_chan_get(channelId, opts): List<Dict>` 


---
_Locked to keep the summary visible. You can always [chat](https://neovim.io/#chat) or open an issue if you have new information/topics to discuss._
